Adding digi-based shower tagging to reco::Muon

battibass opened the pull request:
#### PR description:

This PR adds to the muon object a variable that counts the number of digis per station that are collected in the vicinity of an extrapolated track. Based on this variable, member functions are added to `reco::Muon` to tag showers based on digi information similarly to what is done , for example, [in this study](https://indico.cern.ch/event/804955/#2-follow-up-shower-tagging-met). 

Besides the addition of one more variable to `MuonChamberMatch` (which propagates up to miniAOD), no change to `reco::Muon` objects expected.
 
#### PR validation:

The consistency of the `reco::Muon::nDigisInStation` method with the digi counting made a posteriori in ntuples (used for studies like the ones linked above) has been tested on 1K events of a flat pT muon gun and is available [here](https://battilan.web.cern.ch/battilan/HighPt/PullRequest/hit_count/), in such sample. No discrepancy between the two sources of information was found. 

Event dumps were used to test that `reco::Muon::hasShowerInStation` and `reco::Muon::numberOfShowers` behave as expected.

The increase in event size, for slimmedMuons in miniAOD and for the workflow 136.891, was checked to be less than 1%.

Commit:
DataFormat updates to add digi-based shower info

Commit:
Fix: merge digis from ME1/1a and ME1/1b

Commit:
Add filling of digi-based shower information to MuonIdProducer

Commit:
Remove digi counting in full chamber, leaving only digi counting in |dX| range

Commit:
Add a function that counts showers per muon + cosmetics

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26369/9108

- This PR adds an extra 48KB to repository

cmsbuild commented:
A new Pull Request was created by  @battibass (Carlo Battilana) for master.

It involves the following packages:

DataFormats/MuonReco
RecoMuon/MuonIdentification


@cmsbuild, @perrotta, @slava77 can you please review it and eventually sign? Thanks.
@bellan, @folguera, @abbiendi, @rovere, @jhgoh, @echapon, @calderona, @HuguesBrun, @drkovalskyi, @trocino, @amagitte, @bachtis, @rociovilar this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 4ad4961b6a987247278912eea92e750a4aee821f

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/33998/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests  RelVals**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS


* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
101.0 step1
<pre>runTheMatrix-results/101.0_SingleElectronE120EHCAL+SingleElectronE120EHCAL/step1_SingleElectronE120EHCAL+SingleElectronE120EHCAL.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 4ad4961b6a987247278912eea92e750a4aee821f

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34018/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34018/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3140297
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

smuzaffar commented:
please test
cms-bot PR https://github.com/cms-sw/cms-bot/pull/1115 should fix this

cmsbuild commented:
The tests are being triggered in jenkins.

fabiocos commented:
@smuzaffar I think that the latest failure is unrelated to the issue with the bot

cmsbuild commented:
-1

Tested at: 4ad4961b6a987247278912eea92e750a4aee821f

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34030/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34030/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3140297
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

slava77 commented:
the error in the unit test is :

```
   [5] Calling method for module MuonIdProducer/'muonsFromCosmics'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: MuonDigiCollection<DTLayerId,DTDigi>
Looking for module label: muonDTDigis
Looking for productInstanceName: 

```

this looks related to this PR.

Commit:
Disable muon shower info filling in tau MC embedding merge step

cmsbuild commented:
The code-checks are being triggered in jenkins.

battibass commented:
Hi, the problem with `testTauEmbeddingProducers` should be fixed now.
As in `TauAnalysis/MCEmbeddingTools/python/customisers.py` the filling of other shower-related information in muon objects was already disabled (`process.muons.FillShoweringInfo = cms.bool(False)`), I simply switched off the DIGI-based filling of shower information as well.
Please let me know if this is OK.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26369/9172

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #26369 was updated. @perrotta, @civanch, @mdhildreth, @cmsbuild, @slava77, @santocch can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d88053bbcd57792a4c0307da5cbd736f455b858b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34094/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34094/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3140297
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

civanch commented:
+1

slava77 commented on DataFormats/MuonReco/interface/Muon.h:
```diff
@@ -260,7 +260,17 @@ namespace reco {
+    /// # of digis in a given station layer
+    /// index 0-1-2-3 = DT stations 1-2-3-4                                                             
+    /// index 4-5-6-7 = CSC stations 1-2-3-4
+    int nDigisInStation( int index) const;
```
is there a plan to add RPC or GEMs?
perhaps a function with arguments ```MuonSubdetId, int station``` is more practical.

battibass replied:
I would personally not add RPC or GEM. One can disentangle the signature a shower w.r.t. some channel noise much better in a detector with many layers. My only doubt would be ME0 let me briefly discuss with MUO about this.
Anyhow OK, I will switch to a version getting `MuonSubdetId` as argument in any case.

slava77 commented on RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc:
```diff
@@ -876,6 +888,14 @@ void MuonIdProducer::fillMuonId(edm::Event& iEvent, const edm::EventSetup& iSetu
+       theShowerDigiFiller_->fill(matchedChamber);
+     }
+     else {
+       matchedChamber.nDigisInRange = 0;
```
The pattern above looks like you want to encapsulate setting of potentially more than one parameter; if that's the case,
perhaps ```theShowerDigiFiller_->fillDefault``` is more appropriate.
If there is no plan to add more variables, and only nDigisInRange is going to be computed, then instead of a ```fill(matchedChamber)`` there better be just nDigisInRange computation

battibass replied:
In the offline studies linked as slides at the beginning of the PR I had 4 ranges. I initially implemented this code thinking about 2 ranges, then I switched to one for simplicity. Let me briefly discuss with MUO and, depending on the outcome of the discussion I will change things accordingly.

battibass replied:
I discussed with @Fedespring and we agreed that, depending on the development of the studies being performed for MUO-17-001, the plan is very likely to extend this shower filler (potentially also replacing `MuonShowerInformationFiller`), hence I adopted the solution of adding a `theShowerDigiFiller_->fillDefault()` method.

slava77 commented on RecoMuon/MuonIdentification/src/MuonShowerDigiFiller.cc:
```diff
@@ -0,0 +1,153 @@
+
+{
+
+  edm::InputTag label = iConfig.getParameter<edm::InputTag>("dtDigiCollectionLabel");
```
```Label``` could be omitted and this line can also be joined with the next without a named temporary ```label`` variable.

battibass replied:
OK

slava77 commented on RecoMuon/MuonIdentification/src/MuonShowerDigiFiller.cc:
```diff
@@ -0,0 +1,153 @@
+MuonShowerDigiFiller::~MuonShowerDigiFiller()
+{
+
+}
```
there is no need to define an empty destructor

battibass replied:
OK

slava77 commented on RecoMuon/MuonIdentification/src/MuonShowerDigiFiller.cc:
```diff
@@ -0,0 +1,153 @@
+      double xTrack = muChMatch.x;
+      double yTrack = muChMatch.y;
+
+      for (int iLayer = 1; iLayer < 7; ++iLayer) 
```
```<= CSCDetId::MAX_LAYER``` instead of the magic numbers

battibass replied:
OK

slava77 commented on RecoMuon/MuonIdentification/src/MuonShowerDigiFiller.cc:
```diff
@@ -0,0 +1,153 @@
+      
+      for (int sl = 1; sl < 4; sl += 2) 
+	{
+	  for (int layer = 1; layer < 5; ++layer) 
```
please use ```DTChamberId::maxLayerId ``` and ```maxSuperLayerId``` instead of the magic numbers

battibass replied:
OK

slava77 commented on RecoMuon/MuonIdentification/src/MuonShowerDigiFiller.cc:
```diff
@@ -0,0 +1,153 @@
+	      std::vector<int> adcVals = digiIt->getADCCounts();
+	      bool hasFired = false;
+	      float pedestal = 0.5*(float)(adcVals[0]+adcVals[1]);
+	      float threshold = 13.3 ;
```
should this be configurable as other selections?
If 13.3 has an origin in some calibrations, better use from the source

battibass replied:
I got this from CSC DPG, so I have no answer, let me check with them.

ptcox replied:
Yes, this is historical and arbitrary. It's the signal-above-pedestal threshold considered to be an actual hit rather than a pedestal fluctuation.  It's hard-coded as 13.3 in several places in CSCValidation and CSCOfflineMonitor (part of CSC DQM). This is adc counts, and a very small number relative to the dynamic range of a CSC adc channel (maximum - i.e. saturation - at 4095 counts.)
Should really be configurable because seeing it makes me realize this is something we might like to adjust if pedestal noise increases at higher luminosities (but that's irrelevant for Carlo's purposes here.)

slava77 commented:
@battibass 
please clarify on the status of updates in this PR following the code review.
Thank you.

battibass commented:
@slava77 I will push an update by the end of the day (running one last test now).

Commit:
Follow-up from PR request:
- Use MuonSubdetId in Muon object member functions
- Add fillDefault metshod to MuonShowerDigiFiller
- few other technical changes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26369/9285

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #26369 was updated. @perrotta, @civanch, @mdhildreth, @cmsbuild, @slava77, @santocch can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d9a44d0d30e747926cbccefd992102f7451dc000
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34212/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34212/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3142585
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

civanch commented:
+1

slava77 commented on DataFormats/MuonReco/src/Muon.cc:
```diff
@@ -334,6 +334,78 @@ unsigned int Muon::stationGapMaskPull( float sigmaCut ) const
+bool Muon::hasShowerInStation( int station, int muonSubdetId, int nDtDigisCut, int nCscDigisCut ) const
+{
+  bool hasShower = muonSubdetId == MuonSubdetId::DT ? 
+    nDigisInStation(station,muonSubdetId) >= nDtDigisCut :
```
```c++
if (muonSubdetId != MuonSubdetId::DT || muonSubdetId != MuonSubdetId::CSC) return false;
auto cut = muonSubdetId == MuonSubdetId::DT ? nDtDigisCut : nCscDigisCut;
return nDigisInStation(station,muonSubdetId) >= cut;
```
is perhaps shorter and more explicit about DT vs CSC.

battibass replied:
Right.

battibass replied:
But the OR should be an AND.

slava77 commented on DataFormats/MuonReco/src/Muon.cc:
```diff
@@ -334,6 +334,78 @@ unsigned int Muon::stationGapMaskPull( float sigmaCut ) const
+int Muon::numberOfShowers( int nDtDigisCut, int nCscDigisCut ) const
+{
+  int nShowers = 0;
+  for ( int iCh = 1; iCh < 5; ++iCh )
```
to reduce possible confusion when reading this, perhaps ```iSt``` instead of ```iCh``` because it's supposed to be a shorthand for a station

battibass replied:
Changed in `station` (more consistent with the member functions above).

Commit:
More clear syntax for hasShowerInStation and numberOfShowers

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26369/9301

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #26369 was updated. @perrotta, @civanch, @mdhildreth, @cmsbuild, @slava77, @santocch can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

Commit:
Fix: OR should be AND

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26369/9302

- This PR adds an extra 56KB to repository

cmsbuild commented:
Pull request #26369 was updated. @perrotta, @civanch, @mdhildreth, @cmsbuild, @slava77, @santocch can you please check and sign again.

slava77 commented:
@cmsbuild please abort test

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
@smuzaffar @mrodozov 

I had a request to abort  in https://github.com/cms-sw/cmssw/pull/26369#issuecomment-483736071 ,  which apparently was ignored.
Is this a normal behavior?

cmsbuild commented:
-1

Tested at: d719e62e8541b3e100f5a981de0b42fdf189fb8c

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34230/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests:

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34230/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3142585
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/26369	d719e62
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences
    - I could not find a problem in the unitTest log file and assume that the error reported in https://github.com/cms-sw/cmssw/pull/26369#issuecomment-483887426 is unrelated; the earlier version of this PR (very similar in content) was tested without failures in https://github.com/cms-sw/cmssw/pull/26369#issuecomment-483473744
    - I added a basic plot for ```muMatches_.nDigisInRange```, which should show up in the next update of the reco comparisons for the bot soon.

fabiocos commented:
@mrodozov it looks like the bot is stuck in showing a unit test failure that is no more there, ideas?

davidlange6 commented:
its not a problem of the bot being stuck - it thinks there is a unit test failure. Perhaps one should just restart the tests?

fabiocos commented:
@civanch @santocch could you please have a look?

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

fabiocos commented:
@davidlange6 the point is that I do not see any failure in the report, let's try it again...

mrodozov commented:
@slava77 now I see your question, only "abort" or "please abort" should do. from time to time we have this github webhook failures and that might be it, but I'll assume the bot got confused from your command

mrodozov commented:
@fabiocos I'm searching in the logs and it says 
```
00:11:09 ++ grep -i 'had errors\|recipe for target' /build/cmsbld/jenkins/workspace/ib-any-integration/unitTests.log
00:11:09 ++ sed 's|'\''||g;s|.*recipe for target *||;s|.*unittests_|---> test |;s| failed$| timeout|'
00:11:09 + TEST_ERRORS=
00:11:09 ++ grep -i 'had errors' /build/cmsbld/jenkins/workspace/ib-any-integration/unitTests.log
00:11:09 + TEST_ERRORS=
00:11:09 + true
00:11:09 ++ grep ALL_OK /build/cmsbld/jenkins/workspace/ib-any-integration/unitTests.log
00:11:09 + GENERAL_ERRORS=
00:11:09 + true
00:11:09 Errors in the unit tests
00:11:09 + '[' X '!=' X -o X = X ']'
00:11:09 + echo 'Errors in the unit tests'
00:11:09 + echo 'UNIT_TEST_RESULTS;ERROR'
00:11:09 + ALL_OK=false
00:11:09 + UNIT_TESTS_OK=false
```
which in fact is not true, there are no test failing. 
this part :
```
TEST_ERRORS=
00:11:09 ++ grep -i 'had errors' /build/cmsbld/jenkins/workspace/ib-any-integration/unitTests.log
00:11:09 + TEST_ERRORS=
00:11:09 + true
```
that sets the TEST_ERRORS to true doesn't make any sense, greps for any error, there is none, yet still returns true. lets see if it persists.

fabiocos commented:
@civanch @santocch the SIM and ANA part of this PR is restricted to the Tau embedding configuration where a new flag introduced is set to False by default...

cmsbuild commented:
+1
Tested at: d719e62e8541b3e100f5a981de0b42fdf189fb8c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34272/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26369/34272/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-26369/29034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D41+RecoFullGlobal_2023D41+HARVESTFullGlobal_2023D41

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3213660
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3213455
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

fabiocos commented:
@civanch given the latest changes I will take your previous signature as valid, anyway please check and sign it again for future reference, or comment in case

fabiocos commented:
+1

fabiocos commented:
merge

fabiocos commented on RecoMuon/MuonIdentification/python/MuonShowerDigiFiller_cfi.py:
```diff
@@ -0,0 +1,11 @@
+  ShowerDigiFillerParameters = cms.PSet(
+      digiMaxDistanceX = cms.double(25.0),
+      dtDigiCollectionLabel  = cms.InputTag("muonDTDigis"),
+      cscDigiCollectionLabel = cms.InputTag("muonCSCDigis","MuonCSCStripDigi")
```
@battibass @slava77 it looks like this line is creating issues in wf 1102.0 step4 RECOFROMRECO:

```
   [7] Prefetching for module TrackProducer/'mergedDuplicateTracks'
   [8] Prefetching for module DuplicateTrackMerger/'duplicateTrackCandidates'
   [9] Prefetching for module TrackCollectionMerger/'preDuplicateMergingGeneralTracks'
   [10] Prefetching for module TrackProducer/'muonSeededTracksInOut'
   [11] Prefetching for module CkfTrackCandidateMaker/'muonSeededTrackCandidatesInOut'
   [12] Prefetching for module MuonReSeeder/'muonSeededSeedsInOut'
   [13] Calling method for module MuonIdProducer/'earlyMuons'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: MuonDigiCollection&lt;CSCDetId,CSCStripDigi&gt;
Looking for module label: muonCSCDigis
Looking for productInstanceName: MuonCSCStripDigi

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```

This clearly needs to be fixed, my question is: if this is not affecting any other standard workflow as it seems, and it is not preventing the basic validation of pre4, can we consider to postpone its fix after pre4 is built? Or do you see further possible issues?

Switching this module off in the configuration allows also the step4 of 1102 to run without problems.

battibass replied:
@fabiocos , @slava77 can surely reply more accurately than I can. 
In principle this PR implies that digis are available (hence implies that unpacking was done starting from RAW). For any workflow for which this is the case there should be no problem.
The question is, are there RelVal workflows where this is not the case?
I'm happy to work on a patch asap, but I would need inputs to understand better how I should be testing it, as my check with `runTheMatrix.py -l limited -i all` clearly failed to spot this issue, as well as the one with `TauAnalysis/MCEmbeddingTools/` that was instead spot during integration ...

fabiocos replied:
@battibass indeed this is a bit special case, and only wf 1102.0 implementing it shows this issue. I see that @slava77 has already mentioned it in https://github.com/cms-sw/cmssw/issues/26500 .

slava77 commented:
On 4/19/19 4:18 AM, Carlo Battilana wrote:
> I'm happy to work on a patch asap, but I would need inputs to understand
> better how I should be testing it, as my check with |runTheMatrix.py -l
> limited -i all| clearly failed to spot this issue, as well as the one
> with |TauAnalysis/MCEmbeddingTools/| that was instead spot during
> integration ...

just rewording the earlier message from @fabiocos

runTheMatrix.py -l 1102.0

and then look for the problem in step4


The error does imply that this PR violated assumptions of RECO sequences
design, which apparently implies that digis are not used directly and
local reco outputs are enough to make everything downstream.

Please remind me the logic why it's the digis and not the calibrated
data (which supposedly can also remove the noisy channels)
dt1DRecHits and csc2DRecHits that are used?

battibass commented:
@slava77 I understand well that I can run `runTheMatrix.py -l 1102.0` my question was more general: is there a wider set of tests that I should perform? If so which one?

In principle, one could use dt1DRecHits, for the DTs. They come with left/right ambiguities (there are actually 2 recHits per digi), but it can be handled. For CSC I think the situation is more complex. From a discussion I had with @ptcox when starting all this, the interpretation of csc2DRecHits is not trivial as both clustering and combination of wire and strip digi information happens at RecHit building, and it is not straightforward to understand what the counting of csc2DRecHits actually means for cases beyond the ones where you have clear muon track segments. The use of CSC strip digis was in fact chosen as it should be a better proxy for "counting" the activity in a chamber in case a shower occur.

slava77 commented:
as discussed in the RECO meeting, we can have a quick solution to simply add the CSC and DT digis to the RECO event content.
If there are other reasons to wait with the build of pre4, perhaps this solution can be provided today.

A better choice perhaps is to allow modifications to module PSets of the RECOfromRECO workflows (which is not possible now). See #26500.
We better complete this on the time scale of pre5.

santocch commented:
+1
